python/ray/data/_internal/execution/resource_manager.py

```diff
@@ -224,55 +269,119 @@ def __init__(self, resource_manager: ResourceManager, reservation_ratio: float):
         super().__init__(resource_manager)
         self._reservation_ratio = reservation_ratio
         assert 0.0 <= self._reservation_ratio <= 1.0
-        # We only limit map operators.
-        self._eligible_ops = [
-            op for op in self._resource_manager._topology if isinstance(op, MapOperator)
-        ]
         # Per-op reserved resources.
         self._op_reserved: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Memory reserved for the outputs of each operator. This includes the operator's
+        # internal and external output buffers and next operator's input buffers, but
+        # doesn't include the pending task outputs.
+        # Note, if we don't reserve memory for outputs, all the budget may be used by
+        # by the pending task outputs. Then we'll have no budget to pull the outputs
+        # from the running tasks.
+        self._reserved_for_op_outputs: Dict[PhysicalOperator, int] = {}
         # Total shared resources.
         self._total_shared = ExecutionResources.zero()
-        # Resource limits for each operator.
-        self._op_limits: Dict[PhysicalOperator, ExecutionResources] = {}
+        # Resource budgets for each operator.
+        self._op_budgets: Dict[PhysicalOperator, ExecutionResources] = {}
+
         self._cached_global_limits = ExecutionResources.zero()
+        self._cached_num_eligible_ops = 0
+
+    def _get_eligible_ops(self) -> List[PhysicalOperator]:
+        # Only consider map operators that are not completed.
+        return [
+            op
+            for op in self._resource_manager._topology
+            if isinstance(op, MapOperator) and not op.completed()
+        ]
+
+    def _update_reservation(self):
+        global_limits = self._resource_manager.get_global_limits()
+        eligible_ops = self._get_eligible_ops()
 
-    def _on_global_limits_updated(self, global_limits: ExecutionResources):
-        if len(self._eligible_ops) == 0:
+        if (
+            global_limits == self._cached_global_limits
+            and len(eligible_ops) == self._cached_num_eligible_ops
+        ):
             return
+        self._cached_global_limits = global_limits
+        self._cached_num_eligible_ops = len(eligible_ops)
```

Comment:
Any reason we can't do this in the constructor?

Reply:
When global_limits change or some ops finish, we need to re-calculate them